### ISSUE TITLE
Is the current implementation for requesting a market OK for issuing concurrent calls?

### ISSUE THREAD

**Issue by MeoMix** [etherdelta/etherdelta.github.io#249]
Heyas.

I've made pretty great progress on rewriting all of your example code, but I've got a concern about how I've rewritten logic for capturing market information. Perhaps I've overcomplicated it.

In your logic, I see a couple of concerning things. Let's take a look at the method in question:

```
  self.waitForMarket = (token, user) => new Promise((resolve, reject) => {
    setTimeout(() => {
      reject('Could not get market');
    }, 20000);
    self.state = {
      orders: undefined,
      trades: undefined,
      myOrders: undefined,
      myTrades: undefined,
    };
    self.socket.off('orders');
    self.socket.off('trades');
    const getMarketAndWait = () => {
      self.getMarket(token, user);
      self.socket.once('market', (market) => {
        if ((market.myTrades || !user.addr) && (market.orders && market.trades)) {
          self.updateOrders(market.orders, token, user);
          self.updateTrades(market.trades, token, user);
          self.socket.on('orders', (orders) => {
            self.updateOrders(orders, token, user);
          });
          self.socket.on('trades', (trades) => {
            self.updateTrades(trades, token, user);
          });
          resolve();
        } else {
          setTimeout(() => {
            getMarketAndWait();
          }, 2000);
        }
      });
    };
    getMarketAndWait();
  });
```

My concerns are:

- It's not clear to me, at all, why we need to safeguard against an invalid response from the server. Surely if I've requested information then the server should respond accurately, or emit an error of some kind?

- What if I issue a second request for market information before I've heard back from the server on the first request? It would seem like there's no order guarantee for responses. You might have the second request's response already cached, but need to fetch new data for the first request. This would result in the requests de-synced from the responses and I think that would get our wires crossed up, yeah?

I took a more long-winded approach to it... some parts of this promise management are still wonky (as the comment notes) but you should get the overall idea:

```
  this.pendingMarkets = new Map();

  async createMarket({ symbol, address }){
    console.log(`Creating market: ${symbol}`);
    const market = await (new Promise((resolve, reject) => {
      setTimeout(() => reject('Could not get market'), 20000);

      console.log(`Requesting market information`);
      this.pendingMarkets.set(address, { resolve, reject });
      this.socket.emit('getMarket', { token: address });   
    }));

    return market;
  }

  onSocketMarket({ orders: { sells, buys } } = {}){
    console.log('onSocketMarket');
    forIn(this._getOrderGroups(sells, buys), this.createPendingMarket.bind(this));
  }

  _getOrderGroups(rawSells, rawBuys){
    const sellDtos = map(rawSells, rawSell => new OrderDto(assign({}, rawSell, {
      type: OrderType.Sell, 
      tokenGet: tokenMap.get(rawSell.tokenGet),
      tokenGive: tokenMap.get(rawSell.tokenGive)
    })));

    const buyDtos = map(rawBuys, rawBuy => new OrderDto(assign({}, rawBuy, {
      type: OrderType.Buy, 
      tokenGet: tokenMap.get(rawBuy.tokenGet),
      tokenGive: tokenMap.get(rawBuy.tokenGive)
    })));
    
    const orders = invokeMap(sellDtos.concat(buyDtos), 'toDomain');

    return mapKeys(groupBy(orders, 'token.address'), (orders, tokenAddress) => tokenAddress.toLowerCase());
  }

  // TODO: The structure of this is weird. Returning reject/resolve is not the same as returning a Promise.
  // Create a new Market to represent a token.
  // Market information is received via socket, so market creation
  // requests are marked pending while socket response is awaited.
  createPendingMarket(orders, tokenAddress){
    const token = tokenMap.get(tokenAddress);

    if(!token){
      return Promise.reject(new Error(`Invalid token address: ${tokenAddress}`));
    }

    console.log(`Market information received for ${token.symbol}. Creating with ${orders.length} order${orders.length === 1 ? '' : 's'}`);

    // Ensure the socket's response was requested & expected.
    const pending = this.pendingMarkets.get(tokenAddress);
    if(!pending){
      return Promise.reject(new Error(`Expected ${token.symbol} to be pending creation`));
    }

    // Acknowledge the socket's response was received.
    this.pendingMarkets.delete(tokenAddress);

    // Don't overwrite existing market information. Fail if market
    // is requested again after its initial creation.
    const existing = this.markets.get(tokenAddress);
    if(existing){
      return pending.reject(new Error(`Expected market ${token.symbol} not to exist`));
    }

    // Create a new Market representing the token and existing bid/asks.
    const market = new Market({
      token,
      orderBook: new OrderBook(new Map(map(orders, order => [order.id, order])))
    });
    this.markets.set(tokenAddress, market);

    console.log(`Created market ${token.symbol}`);
    return pending.resolve(market);
  }
```

TL;DR: Instead of making assumptions on the markets being returned - just issue a request for market information. Accept all market responses, group the received information by tokenAddress, and create markets with this information.

I think this side-steps any race conditions, but I really dislike making my service stateful with the "pendingMarkets" concept. So, I thought I would get your opinion on whether it's overkill.

**Comment by MeoMix** [comment]
I eventually decided I liked your approach to things more than mine. Current code looks something like:

```js
  // Get current order book information, represented as a Market, for the given token.
  // EtherDelta's socket does not always return expected information on first try. So,
  // allow for retrying.
  getMarket(token){
    logger.info(`Getting market: ${token.symbol}`);

    return new Promise((resolve, reject) => {
      const onSocketMarketResponse = async response => {
        // It's somewhat common for EtherDelta's socket to respond with an empty orders object.
        // Retry automatically when given a bad response.
        if(!response || !response.orders){
          logger.warn(`Market response did not include orders. Reissuing getMarket request`);
          // Pause before re-issuing the query to prevent spamming the server.
          await sleep(1000);
          this.socket.emit('getMarket', { token: token.address }); 
          return;
        }
        this.socket.off('market', onSocketMarketResponse);
    
        const orderGroups = this._getOrderGroups(response.orders.sells, response.orders.buys);
        if(keysIn(orderGroups).length > 1){
          console.warn(`Expected response to only include one market`);
        }

        const orders = orderGroups[token.address];
        if(!orders){
          reject(new Error(`${token.symbol} is missing from the response`));
          return;
        }

        // Create a new Market representing the token and existing bid/asks.
        logger.info(`${token.symbol} market info received. Creating with ${orders.length} order${orders.length === 1 ? '' : 's'}`);
        resolve(new Market({ token, orderBook: new OrderBook(orders) }));
      };
      this.socket.on('market', onSocketMarketResponse);

      setTimeout(() => {
        this.socket.off('market', onSocketMarketResponse);
        reject(new Error(`Failed to get market for ${token.symbol} within 30 seconds`));
      }, 30000);

      this.socket.emit('getMarket', { token: token.address });  
    });
  }
```
